Implement randomness into active and dormant buying periods in toolkit/buy policy

nuclearkatie opened the pull request:
Closes #1622 and #1641

gonuke commented:
I'm not sure why that conda test is failing... I'm investigating now

gonuke commented:
Now that I fixed the docker image this is passing all tests- happy to consider merging

nuclearkatie commented:
Oops, forgot it was still in draft status @gonuke, please merge or add more review if necessary. Thanks!

gonuke commented on src/random_number_generator.cc:
```diff
@@ -57,7 +57,54 @@ namespace cyclus{
+    UniformDoubleDist::UniformDoubleDist(double min, double max) : dist(min, max) {}
+
+    double UniformDoubleDist::sample() { 
+        return UniformDoubleDist::dist(RandomNumberGenerator::gen_); 
```
Shouldn't this just be:

```suggestion
        return dist(RandomNumberGenerator::gen_); 
```

since `dist` is a member variable of this particular object?

gonuke commented on src/random_number_generator.cc:
```diff
@@ -57,7 +57,54 @@ namespace cyclus{
+    }
+
+    double UniformDoubleDist::max() { 
+        return UniformDoubleDist::dist.max(); 
```
same as above

gonuke reviewed (changes requested):
It seems that you are over-specifying the `dist` in many of your routines here?  I put a couple of comments, but not sure if I'm missing something.

gonuke commented:
> something from PR #1640 seems to have prevented the CI from being able to install dependencies

I think that's my fault for suggesting the `Dockerfile` change to force `libxmlpp-4.0` in this branch.  Any changes to the `Dockerfile` force it to try and rebuild the CI images, but the CI images can only be rebuilt from a branch of the upstream repo.

If you rebase on `main` that `Dockerfile` change will go away and then it should pass all tests.

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -55,34 +53,45 @@ void MatlBuyPolicy::set_throughput(double x) {
+  active_dist_ = new FixedIntDist(1);
+  dormant_dist_ = new FixedIntDist(-1);
+  size_dist_ = new FixedDoubleDist(1.0);
+  init_active_dormant();
```
Since this gets called explicitly in cases where you don't use default distributions, I think it helps readability to call it explicitly after calling `set_default_distributions` in all the places where you do use the defaults.

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -55,34 +53,45 @@ void MatlBuyPolicy::set_throughput(double x) {
-  assert(x >= 0);
-  dormant_ = x;
+void MatlBuyPolicy::init_active_dormant() {
+  next_active_end_ = active_dist_->sample();
```
Doesn't this need to be added to `manager()->context()->time()` in case this policy gets initialized after time 0?

Or just call `SetNextActiveTime()`

gonuke replied:
What happens if `next_active_end_ < 0`?

nuclearkatie replied:
If I add a call to `manager()->context()->time()` here I get a seg fault when building

gonuke replied:
It's odd that it seg faults when building???  I don't have the fine details of the order of initializing everything in my head, but conceptually, I could imagine a seg fault if an agent invokes this initialization before the context has initialized time - I'd have to dig to see if that's what's happening and then figure out what the best path forward would be.  I think the agent has to exist (?? and been added to the context ??) before you `Init()` the buy policy.

In general, it seems that the first time that `next_active_end_` is set does need to be relative to the time at which the agent enters the simulation.  Calling `SetNextActiveTime()` should take care of that (but it may also result in a segfault?)

Our tests may not catch this and may always have agents enter at time 0 - if that's true, it's not your problem to fix as part of this PR, but we should make an issue to improve testing.

nuclearkatie replied:
👍 added check of dist max in init, throw ValueError if >1. Removed check in SampleRequestSize

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -199,10 +226,32 @@ void MatlBuyPolicy::AcceptMatlTrades(
+void MatlBuyPolicy::SetNextDormantTime() {
+  if (next_dormant_end_ < 0) {}
+  else {
+    next_dormant_end_ = dormant_dist_->sample() + manager()->context()->time();
```
doesn't this need to be
```suggestion
    next_dormant_end_ = dormant_dist_->sample() + next_active_end_;
```

nuclearkatie replied:
I think these are functionally equivalent? `SetNextDormantTime` is only called at the end of an active period, so `manager()->context()->time()` is the same as `next_active_end_`. If you think using `next_active_end_` is better, 👍

gonuke commented on src/toolkit/matl_buy_policy.h:
```diff
@@ -128,15 +141,18 @@ class MatlBuyPolicy : public Trader {
 
+  double sample_fraction_ = 1.0;
+
   /// the total amount requested
   inline double TotalQty() const {
-    return std::min(throughput_,
-                    fill_to_ * buf_->capacity() - buf_->quantity());
+    return std::min((throughput_* sample_fraction_),
+                    ((fill_to_ * buf_->capacity()) - buf_->quantity()) * sample_fraction_);
```
Since 
* `sample_fraction` is only needed as part of `TotalQty()`, 
* it is set by `SetSampleSize()`, and
* `SetSampleSize()` must be called before `TotalQty()` and only then

why not just add the code from `SetSampleSize()` to `TotalQty()` and get rid of the former and the member variable.

gonuke replied:
Also, perhaps multiply the result of the `std::min()` by `sample_fraction_` instead of multiplying both and then taking the min?

nuclearkatie replied:
That's not exactly true. Yes, , `SetRequestSize()` is only called immediately before `TotalQty()` in `matl_buy_policy.cc` directly, but `TotalQty()` is also called through member functions `ReqQty()` or `NReq()`, which also happens in the same time step. If I move `sample_fraction` to be part of `TotalQty()`, then `sample_fraction` will be re-set multiple times in a single timestep, because it calls a new sample every time

gonuke commented on src/toolkit/matl_buy_policy.h:
```diff
@@ -173,13 +193,22 @@ class MatlBuyPolicy : public Trader {
   double fill_to_, req_when_under_, quantize_, throughput_;
-  int active_, dormant_;
+  
+  double cumulative_capacity_;
```
This isn't used.

nuclearkatie replied:
Oops, left over from thinking about adding cumulative capacity (which is the next step)

gonuke commented on src/toolkit/matl_buy_policy.h:
```diff
@@ -85,15 +86,27 @@ class MatlBuyPolicy : public Trader {
   MatlBuyPolicy& Init(Agent* manager, ResBuf<Material>* buf, std::string name,
-                      double throughput, int active = 1, int dormant = 0);
+                      double throughput, 
+                      IntDistribution* active_dist = new FixedIntDist(1),
+                      IntDistribution* dormant_dist = new FixedIntDist(0),
+                      DoubleDistribution* size_dist = new FixedDoubleDist(1.0));
```
This may create a memory leak if these are not deleted somewhere.  However, we can't **always** delete them since they may be passed in from somewhere else and we don't have the "authority" to delete them.

I think the best solution is to use `NULL` as the default and then instantiate these in the constructor if the value is `NULL`.  Then, you'll also want these variables to use `boost::shared_ptr<>`.  This should automatically track whether there continue to be references to the objects in question.

nuclearkatie replied:
But the defaults aren't overwritten until init, so the values would always be null at construction

gonuke replied:
Oops! I keep getting `Init()` confused with the constructor... let me rephrase:

I think the best solution is to use `NULL` as the default and then instantiate these in `Init()` if the value is NULL. Then, you'll also want these variables to use boost::shared_ptr<>. This should automatically track whether there continue to be references to the objects in question.

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -153,24 +164,40 @@ std::set<RequestPortfolio<Material>::Ptr> MatlBuyPolicy::GetMatlRequests() {
+
+  int current_time_ = manager()->context()->time();
+
+  if (current_time_ < next_active_end_ || next_dormant_end_ < 0) {
```
A little inline private function might make this more readable:
```
bool never_dormant() { return next_dormant_end_ < 0;};
```

gonuke reviewed (changes requested):
(I'm not done reviewing but need to move on to something else, so I'm just submitting these for now.)

I need to think through the usage of `next_active_end_` and `next_dormant_end_` to see if there is something simpler...

nuclearkatie commented on src/random_number_generator.cc:
```diff
@@ -57,7 +57,54 @@ namespace cyclus{
+    }
+
+    NormalIntDist::NormalIntDist(double mean, double std_dev, int min, int max) : dist(mean, std_dev) {
+        min_ = std::min(min, 0);
```
Guess I was only thinking of applying distributions to things like time or size, which have physical constraints. Do you think there's an obvious use case for negative numbers?

gonuke replied:
I don't think there is anything obvious, but I'm not sure there is a reason to disallow it.

gonuke commented on docker/Dockerfile:
I think this change will collapse when you rebase on the current main branch, and then it won't try to do the CI-image-building tests that are failing.

gonuke commented on src/random_number_generator.cc:
```diff
@@ -57,7 +57,38 @@ namespace cyclus{
+        return rounded_val;
+    }
+
+    //
+    // Distributions
+    NormalDoubleDist::NormalDoubleDist(double mean, double std_dev, double min, double max) : dist(mean, std_dev) {
+        min_ = min;
+        max_ = max;
+    }
```
This can probably be inlined in the header.

gonuke commented on src/random_number_generator.cc:
```diff
@@ -57,7 +57,38 @@ namespace cyclus{
         return val;
     }
 
+    double NormalDoubleDist::max() { 
+        return dist.max();
+    }
```
This can be inlined in the header

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -55,34 +53,46 @@ void MatlBuyPolicy::set_throughput(double x) {
+  if (dormant_len < 0) {
+    next_dormant_end_ = -1;
+  }
+  else {
+    next_dormant_end_ = dormant_len + next_active_end_;
+  }
```
Assuming it makes sense to call `SetNextActiveEnd()` above, it will make this more readable if you just call `SetNextDormandEnd()` here as well.  As long as you set `next_dormant_end_` to be -1 before then, when appropriate, it will result in the same behavior, but waste a single random number.

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -153,24 +167,40 @@ std::set<RequestPortfolio<Material>::Ptr> MatlBuyPolicy::GetMatlRequests() {
+    SetRequestSize();
     amt = TotalQty();
   }
-  else if (dormant_ == 0)
-    amt = TotalQty();
-  else {
-    // in dormant part of cycle, return empty portfolio
+  else if (current_time_ < next_dormant_end_) {
+    // currently in the middle of dormant period
     amt = 0;
-    LGH(INFO3) << "in dormant period, no request";
+    LGH(INFO3) << "in dormant period, no request" << std::endl;
+  }
+  else if (current_time_ == next_active_end_) {
+    // finished active. starting dormancy and sample/set length of dormant period
+    amt = 0;
+    SetNextDormantTime();
+  }
+  // the following is an if rather than if-else statement because if dormant
+  // length is zero, buy policy should return to active immediately
+  if (current_time_ == next_dormant_end_) {
+    // finished dormant. starting buying and sample/set length of active period
+    SetRequestSize();
+    amt = TotalQty();
+    SetNextActiveTime();
   }
```
This section may be more readable as (pseudo-code):
```
if (!never_dormant() || current_time_ < next_active_end_) {
       do something
}
else if (current_time_ == next_active_end) {
    do other things
}
else if (current_time_ < next_dormant_end_) {
   do still different things
}
else if (current_time_ == next_dormant_end_) {
  do last batch of things
}
```

gonuke commented on src/toolkit/matl_buy_policy.h:
```diff
@@ -128,15 +141,18 @@ class MatlBuyPolicy : public Trader {
 
+  double sample_fraction_ = 1.0;
+
   /// the total amount requested
   inline double TotalQty() const {
-    return std::min(throughput_,
-                    fill_to_ * buf_->capacity() - buf_->quantity());
+    return sample_fraction_ * std::min(throughput_,
+           (fill_to_ * buf_->capacity()) - buf_->quantity());
```
I think you just want to move the logic/code from `SetRequestSize()` into `TotalQty()` since that code is needed if and only if `TotalQty()` is called.

gonuke commented on tests/toolkit/matl_buy_policy_tests.cc:
```diff
@@ -181,12 +182,49 @@ TEST_F(MatlBuyPolicyTests, MultiReqQuantize) {
-  int active = 2;
-  int dormant = 3;
-  int dur = 4;
+TEST_F(MatlBuyPolicyTests, DefaultFixedActiveDormant) {
+  using cyclus::QueryResult;
+  
+  FixedIntDist a_dist = FixedIntDist(1);
+  FixedIntDist d_dist = FixedIntDist(-1);
+  FixedDoubleDist size_dist = FixedDoubleDist(1.0);
```
Doesn't it make sense for this test to call `Init()` with no values for these and rely on the defaults defined in the header?  That will test both:

- that the defaults result in the default behavior
- the default behavior is to have no dormant phase

gonuke commented on tests/toolkit/matl_buy_policy_tests.cc:
```diff
@@ -181,12 +182,49 @@ TEST_F(MatlBuyPolicyTests, MultiReqQuantize) {
+TEST_F(MatlBuyPolicyTests, FixedActiveDormant) {
+  using cyclus::QueryResult;
+  
+  FixedIntDist a_dist = FixedIntDist(1);
+  FixedIntDist d_dist = FixedIntDist(1);
```
I'm not sure what our best practice should be, but you could re-use these distributions if their identical, especially if they are `FixedIntDist`s
```suggestion
  FixedIntDist a_d_dist = FixedIntDist(1);
```
and pass `a_d_dist` for both the active and dormant distribution.

gonuke commented on tests/toolkit/matl_buy_policy_tests.cc:
```diff
@@ -235,13 +320,187 @@ TEST_F(MatlBuyPolicyTests, ActiveDormantMultipleCycles) {
+
+  cyclus::toolkit::ResBuf<cyclus::Material> inbuf;
+  cyclus::toolkit::MatlBuyPolicy policy;
+  policy.Init(fac, &inbuf, "inbuf", throughput, &a_dist, &d_dist, &size_dist)
+        .Set("commod1").Start();
```
If your default was `NULL` you could also actively just pass `NULL` to get default behavior.
```suggestion
  policy.Init(fac, &inbuf, "inbuf", throughput, NULL, NULL, &size_dist)
        .Set("commod1").Start();
```

gonuke reviewed (changes requested):
This is looking pretty good.  I think we're nearly there - happy to discuss any of these requests/suggestions.

gonuke commented on src/toolkit/matl_buy_policy.h:
```diff
@@ -128,23 +142,26 @@ class MatlBuyPolicy : public Trader {
   /// is idempotent.
   void Stop();
 
-  /// the total amount requested
-  inline double TotalQty() const {
+  double random_request_size_ = 1.0;
+
+  /// the total amount available to request
+  inline double TotalAvailable() const {
     return std::min(throughput_,
-                    fill_to_ * buf_->capacity() - buf_->quantity());
+           (fill_to_ * buf_->capacity()) - buf_->quantity());
```
The original indentation is probably preferred

gonuke reviewed (changes requested):
I've left a couple of little suggestions here.

gonuke commented on src/random_number_generator.h:
```diff
@@ -76,6 +84,74 @@ class RandomNumberGenerator {
+  public:
+    NormalDoubleDist(double mean, double std_dev, double min=0, double max=std::numeric_limits<double>::max()) : dist(mean, std_dev), min_(min), max_(max) {};
+    virtual double sample();
+    virtual double max() { return dist.max(); }
```
Should this be `dist.max()` or `max_`?

nuclearkatie replied:
ahh, it should be `max_`

gonuke commented on src/random_number_generator.h:
```diff
@@ -76,6 +84,74 @@ class RandomNumberGenerator {
+  public:
+    NormalIntDist(double mean, double std_dev, int min=0, int max=std::numeric_limits<int>::max()) : dist(mean, std_dev), min_(min), max_(max) {};
+    virtual int sample();
+    virtual int max() { return dist.max(); }
```
`dist.max()` or `max_`?

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -153,24 +170,39 @@ std::set<RequestPortfolio<Material>::Ptr> MatlBuyPolicy::GetMatlRequests() {
-    amt = TotalQty();
-  else {
-    // in dormant part of cycle, return empty portfolio
+  else if (current_time_ == next_active_end_) {
+    // finished active. starting dormancy and sample/set length of dormant period
+    amt = 0;
+  }
```
How is this condition different from the next one?  Do you need it?

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -153,24 +170,39 @@ std::set<RequestPortfolio<Material>::Ptr> MatlBuyPolicy::GetMatlRequests() {
+  // length is zero, buy policy should return to active immediately
+  if (current_time_ == next_dormant_end_) {
+    // finished dormant. starting buying and sample/set length of active period
+    SetRequestSize();
+    amt = TotalAvailable() * random_request_size_;
```
Update as above (assuming that `SetRequestSize()` returns a double as commented below)

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -199,10 +231,32 @@ void MatlBuyPolicy::AcceptMatlTrades(
+  return;
+}
+
+void MatlBuyPolicy::SetRequestSize() {
+  random_request_size_ = size_dist_->sample();
+  if (random_request_size_ > 1) {
+    random_request_size_ = random_request_size_ / size_dist_->max();
+  }
+  return;
+}
```
With the above update, this now should return a `double` and we can eliminate the member variable `random_request_size_` in favor of a local variable here.

gonuke commented on tests/toolkit/matl_buy_policy_tests.cc:
```diff
@@ -226,25 +277,239 @@ TEST_F(MatlBuyPolicyTests, ActiveDormantMultipleCycles) {
+  delete a;
+}
+
+// TEST_F(MatlBuyPolicyTests, MixedActiveDormant) {
```
Are these tests commented out on purpose?

nuclearkatie replied:
oops, no. at least, not anymore

gonuke reviewed (changes requested):
(Not sure if this was ready for review... or maybe only partly finished?) but here are some comments

nuclearkatie commented:
@gonuke addressed all comments (I think) in the latest commits.

I think I figured out what's going on with the seg fault in testing (not build, that was my mistake). So, if a buy policy is initialized normally, with a manager, then a timer (and a time, I think) exist at init.

However, I realized that the test that was giving me the seg fault is `StartStop`, which tests for a `ValueError` when a policy is initialized with a NULL manager. NULL manager = no access to `time()` = seg fault.

I addressed this by checking if manager is NULL as a first thing during init (`set_manager`), and throwing a ValueError at that point. Then, I can call `SetNextActiveTime()` and `SetNextDormantTime()` without issue because anything that makes it to `init_active_dormant` should be confirmed to have a non-null manager, and thus be reasonably expected to access `time()` with no difficulty. This does affect previous tests because now you can't get to start (or stop) if you don't have a manager. I assume that's alright?

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -153,24 +184,35 @@ std::set<RequestPortfolio<Material>::Ptr> MatlBuyPolicy::GetMatlRequests() {
-    amt = TotalQty();
-  else {
-    // in dormant part of cycle, return empty portfolio
+  else if (current_time_ == next_active_end_ || current_time_ < next_dormant_end_) {
```
The only way to get here is if `next_dormant_end_ > 0` and `next_active_end_ < current_time < next_dormant_end_ ` so I think you only need to test against `next_dormant_end_`:
```suggestion
  else if (current_time_ < next_dormant_end_) {
```

gonuke commented on src/toolkit/matl_buy_policy.cc:
```diff
@@ -199,10 +241,33 @@ void MatlBuyPolicy::AcceptMatlTrades(
+
+double MatlBuyPolicy::SampleRequestSize() {
+  double random_request_size_ = size_dist_->sample();
+  if (random_request_size_ > 1) {
+    random_request_size_ = random_request_size_ / size_dist_->max();
+  }
```
This seems like it will give inconsistent behavior - if `size_dist_->max() > 1` it's possible to get a random sample that is less than 1 that will NOT be normalized by `size_dist_->max()` and then another sample that is greater than 1 that WILL be normalized.  

Maybe the condition should be to divide by `size_dist_->max()` if `size_dist_->max() > 1`??

nuclearkatie replied:
You know, as I'm thinking about this, it may make sense for all reasonable user use cases (like myself, encountering this while writing the tests). You're right that my original version is inconsistent in any case. Your version makes sense In the context of uniform distribution, which is required to have a max, but less so for normal, which is not. For a normal dist, the max defaults to the numeric limit. Therefore, a default normal dist, which only sets mean and stddev, will constantly be scaled against the largest value possible, resulting in a zero request size. 
Instead of adding fancy handling here, it may be simplest to just say that if 
```suggestion
  if (random_request_size_ > 1) {
    random_request_size_ = 1.0;
  }
```
aka the simplest artificial max

gonuke replied:
Two things:

1) This will skew the sampling of the distribution.  What if someone has a Normal Distribution with a max of 1.1?  The sampling routine will reject/resample until it gets a value less than 1.1, but this will simple change the value rather than resample.  That's an inconsistent handling of your stochastic variable.

2) Therefore, it might make more sense to require all distributions used for sample size to have a max that is less than or equal to 1?  This could be tested when the policy is initialized and cause an error (or automatic correction) when violated.

gonuke commented on tests/toolkit/matl_buy_policy_tests.cc:
```diff
@@ -235,13 +324,182 @@ TEST_F(MatlBuyPolicyTests, ActiveDormantMultipleCycles) {
+  using cyclus::QueryResult;
+
+  boost::shared_ptr<NormalIntDist> a_dist = 
+  boost::shared_ptr<NormalIntDist>(new NormalIntDist(5, 1, 2, 10));
```
indent or combine with previous line

gonuke reviewed (changes requested):
Two minor comments and one more substantive one (`SampleRequestSize()`)

gonuke commented on src/random_number_generator.h:
```diff
@@ -76,6 +86,88 @@ class RandomNumberGenerator {
+        throw ValueError("Min and max cannot be equal for a normal distribution. Either use FixedDoubleDist or change the min/max.");
+      }
+      if (max_ < (mean - 3*std_dev) || min_ > (mean + 3*std_dev)) {
+        Warn<VALUE_WARNING>("Dist is sampling from a truncated normal more than 3 standard deviations from the mean. Drawing sampling may be inefficient");
```
```suggestion
        Warn<VALUE_WARNING>("Dist is sampling from a tail of a truncated normal more than 3 standard deviations from the mean. Drawing sampling may be inefficient");
```
